apps/snitch_core/lib/core/data/schema/user.ex

```diff
@@ -85,23 +88,51 @@ defmodule Snitch.Data.Schema.User do
     |> common_changeset()
   end
 
-  def delete_changeset(user, params) do
+  def delete_changeset(user, _params \\ %{}) do
+    params = %{
+      "state" => "deleted",
+      "deleted_at" => NaiveDateTime.utc_now()
+    }
+
     user
-    |> cast(params, @required_fields)
-    |> no_assoc_constraint(:orders, message: "Cannot delete as orders are associated")
+    |> cast(params, [:state, :deleted_at])
   end
 
   @spec common_changeset(Ecto.Changeset.t()) :: Ecto.Changeset.t()
   defp common_changeset(user_changeset) do
     user_changeset
-    |> unique_constraint(:email, message: "This email is already registered")
+    |> unique_constraint(:email, name: :unique_email)
+    |> unique_email_validation()
     |> foreign_key_constraint(:role_id)
     |> validate_confirmation(:password)
     |> validate_length(:password, min: @password_min_length)
     |> validate_format(:email, ~r/@/)
     |> put_pass_hash()
   end
 
+  defp unique_email_validation(%Ecto.Changeset{} = changeset) do
+    case fetch_change(changeset, :email) do
+      {:ok, email} ->
+        email_user_exists(email, changeset)
+
+      :error ->
+        changeset
+    end
+  end
+
+  defp email_user_exists(email, changeset) do
+    user = from(u in __MODULE__, where: u.state == "active" and u.email == ^email) |> Repo.one()
```

Comment:
@jyotigautam here also we are using strings, we should atoms? thoughts ?

Reply:
Enum allows us to pass an integer, atom or string to it while performing a query on the state(or any field with EctoEnum type) field.